src/rust/engine/fs/src/lib.rs

```diff
@@ -532,6 +532,20 @@ impl PathGlobs {
       })
     }))
   }
+
+  pub fn filter(&self, paths: Vec<PathBuf>) -> Result<Vec<PathBuf>, String> {
+    let mut output = vec![];
+    for path in paths.into_iter() {
+      match self.matches(&[path.clone()]) {
```

Comment:
`self.matches` is implemented as a loop (a filtering iterator) already: I'd suggest directly using its internal implementation (probably by breaking out another method that returns `Result<impl Iterator<..>, String>`) and then calling `collect()`.

Reply:
I ended up making a fairly substantial change to `PathGlobs` to make this workable